### ISSUE TITLE
How to Cross complie libsndfile for arm.

### ISSUE THREAD

**Issue by blprasad** [erikd/libsndfile#207]
Hi All,
I am trying to cross compile linsndfile sources on ubuntu machine.I am getting the following error

/bin/bash: ./create_symbols_file: cannot execute binary file: Exec format error
make[2]: *** [Symbols.gnu-binutils] Error 126
make[2]: Leaving directory `/home/prasad/orangepi/libsndfile/build-orangepi/src'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/prasad/orangepi/libsndfile/build-orangepi/src'
make: *** [all-recursive] Error 1

I googled and found that "create_symbols_file" is not compatible on host machine.
If its the case how could I cross compile successfully on ubuntu host and target is Arm.
Thanks in advance.

**Comment by evpobr** [comment]
Could you try https://github.com/evpobr/libsndfile/tree/cmake-config-revised? There, the procedure for generating symbol files has been significantly changed. If Autotools build  does not work out for you, try using a CMake build system.

**Comment by blprasad** [comment]
Hi,
I am knew to CMake,but I am trying to learn it for the past few days, I tried experimenting with CMake.
But I could not succeed so far,I may need to dig more. I tried on CMake-gui as an attempt but I get an error as CMake 3.0.0 is minimal needed for compiling sorces but CMake-gui on ubuntu uses only 2.8.

Thanks for reply

**Comment by evpobr** [comment]
* What version of libsndfile did you try to build?
* Have you installed all the necessary packages for cross-compilation?
* What parameters did you run configure script with?

**Comment by blprasad** [comment]
Hi evpobr, 

    What version of libsndfile did you try to build?
    It is libsndfile 1.0.27, I knew that form README file of libsndfile sources. 
    
   Have you installed all the necessary packages for cross-compilation?
    I could not understand about necessary packages, do you mean libsndfile need dependencies to be    ready before compiling? Please guide me in this ,if i am missing any. 
   
   What parameters did you run configure script with?
   this is how i am compiling 
  #./configure  --host="arm-linux-gnueabihf" --prefix=/home/prasad/orangepi/zero/gcc-linaro-6.2.1-2016.11-i686_arm-linux-gnueabihf/sysroot/usr --with-sysroot=/home/prasad/orangepi/zero/gcc-linaro-6.2.1-2016.11-i686_arm-linux-gnueabihf/sysroot
   
 I have installed arm-linux-gnueabihf-gcc  toolchain in my system.
 I compiled libsodium,opus, this way so i tried the method with libsndfile sources too.

my objective is to cross compile Toxcore for Arm  ,and libsndfile is one among Toxcore depedencies.
Let me know if  I am missing anything.
 
thanks for your effort and time

**Comment by evpobr** [comment]
> It is libsndfile 1.0.27, I knew that form README file of libsndfile sources. 

from `erikd/master`? It is development branch, far from v1.0.27. Use latest [release](https://github.com/erikd/libsndfile/releases/tag/1.0.27).

It seems like your problem is related to new code, implemented after v1.0.27 in master branch. `create_symbols_file` is bundled utility to generate export symbol files. When cross-compiling, `create_symbols_file` is for ARM CPU and it cannot run on your native CPU. This is bug.

In v1.0.27 symbol files are generated by python script, and it should work anywhere.

Also, if you have time, please update your CMake from backports and check how it works when cross-compiling with [my branch](https://github.com/evpobr/libsndfile/tree/cmake-config-revised).

**Comment by erikd** [comment]
If you check out commit 2fcf531ac9 cross-compiling should work with the autotools build system.

**Comment by evpobr** [comment]
@erikd, it is issue under Linux when cross-compiling for ARM. That commit fixed MSVC issue.

**Comment by erikd** [comment]
Commit 2fcf531ac94 fixed a Windows issue. The commit immediately after it, ae272b59c0a680, broke cross compiling.

**Comment by blprasad** [comment]
Hi evpobr,
as you recommended tried with source from 
https://github.com/erikd/libsndfile/releases/tag/1.0.27

I am getting the following error during make 

/usr/lib/gcc-cross/arm-linux-gnueabihf/4.8/../../../../arm-linux-gnueabihf/bin/ld: cannot open linker script file ../../src/Symbols.gnu-binutils: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [libsndfile.la] Error 1
make[2]: Leaving directory `/home/prasad/orangepi/libsndfile-1.0.27/build-orangepi/src'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/prasad/orangepi/libsndfile-1.0.27/build-orangepi/src'
make: *** [all-recursive] Error 1

and your other recommendation 
Also, if you have time, please update your CMake from backports and check how it works when cross-compiling with my branch.

I could not understand that, do  I need to replace CMake directory of mine with your reference CMake and try?

**Comment by evpobr** [comment]
> #./configure --host="arm-linux-gnueabihf" --prefix=/home/prasad/orangepi/zero/gcc-linaro-6.2.1-2016.11-i686_arm-linux-gnueabihf/sysroot/usr --with-sysroot=/home/prasad/orangepi/zero/gcc-linaro-6.2.1-2016.11-i686_arm-linux-gnueabihf/sysroot

Did you run  `./autogen.sh` before `./configure`?

**Comment by evpobr** [comment]
Got some test results. `master` - fails. 2fcf531 - works.

Here is my command line:

```
sudo apt-get install gcc-arm-linux-gnueabi g++-arm-linux-gnueabi
```

```
./autogen.sh
./configure --host=arm-linux-gnueabi
make
```

```
$ file libsndfile.so.1.0.28
libsndfile.so.1.0.28: ELF 32-bit LSB  shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=12cb4edf547cd4fabfa28f37767480f316a887b2, not stripped
```

CMake from this PR fails.

**Comment by evpobr** [comment]
@erikd: CMake fails with:

```
src/file_io.c: in function 'psf_get_filelen_fd':
src/file_io.c:481:16: error: error: storage size of 'statbuf' isn't known 
src/file_io.c:483:2: warning: implicit declaration of function ‘fstat64’ [-Wimplicit-function-declaration] 
```

Something is not configured properly. Is it about large files support?

**Comment by erikd** [comment]
Yes, almost certainly large file support related.

**Comment by blprasad** [comment]
Hi evpobr,
I used 
#autoreconf -i   prior to ./configure

but now I tried with 
#./autogen.sh      and this gives me the following

./autogen.sh: 187: ./autogen.sh: cannot create .git/hooks/pre-commit: Directory nonexistent
chmod: cannot access ‘.git/hooks/pre-commit’: No such file or directory

**Comment by evpobr** [comment]
@blprasad 

> I used
> #autoreconf -i prior to ./configure

Yes, there is nothing in `README` about building from GitHub releases. But there are some explanations in `README.md`. It says you must use `./autogen.sh`.

OK then, @erikd recommends compile from official tar-ball. Let's try with the latest stable release v1.0.27. Download http://www.mega-nerd.com/libsndfile/files/libsndfile-1.0.27.tar.gz and extract to separate directory. Do standard stuff:

```
./configure <your-cross-compile-options>
make
```

**Comment by blprasad** [comment]
@evpobr 
I did as you suggested ,downloaded tar.gz
I got the following during make

../src/.libs/libsndfile.so: error adding symbols: File in wrong format
collect2: error: ld returned 1 exit status
make[1]: *** [sndfilehandle] Error 1
make[1]: Leaving directory `/home/prasad/orangepi/libsndfile-1.0.27/build-orangepi/examples'
make: *** [all-recursive] Error 1

I am attaching my make file here ,if possible please try to look at it, you may find any suspicious or wrong in it .

[Makefile.zip](https://github.com/erikd/libsndfile/files/857234/Makefile.zip)

**Comment by erikd** [comment]
@blprasad You should try to make your posts as informative as possible being particularly careful not to leave out details. You should also note that providing a 32k machine generated Makefile is of no help to anyone.

On my Debian system I did the following:
```
cd /tmp/
git clone https://github.com/erikd/libsndfile
cd libsndfile/
git checkout 2fcf531 -b cross-compile
./autogen.sh 
./configure --host=arm-linux-gnueabi
make
```
The library built fine, but `sndfiile-play` didn't build because of a missing library:
```
make[1]: Entering directory '/tmp/libsndfile/programs'
  CCLD     sndfile-play
/usr/lib/gcc-cross/arm-linux-gnueabi/6/../../../../arm-linux-gnueabi/bin/ld: cannot 
find -lasound
```
However, like I said, the library was fine:
```
> file src/.libs/libsndfile.so.1.0.28 
src/.libs/libsndfile.so.1.0.28: ELF 32-bit LSB shared object, ARM, EABI5 version 1 
(SYSV), dynamically linked, BuildID[sha1]=75753817951e55a28aa0264f84401f3e2945b5e1, 
not stripped
```

**Comment by blprasad** [comment]
@erikd 

Sorry ,if I have given you any trouble.Why I send you Makefile,is just as a request.
I observed the following in my Makefile I felt it bit suspecious as there is "unknown" string
target = arm-unknown-linux-gnueabihf, I don't know if its Ok,"unknown" string appears at several places

And another thing is ,I don't know if I could use 
./configure --host=arm-linux-gnueabi

I doubt to use
arm-linux-gnueabi-gcc   because I am confused between arm-linux-gnueabi and arm-linux-gnueabihf
here someone also raised the same doubt
http://nairobi-embedded.org/050_linaro_arm_toolchain.html

right now I am using the tool chain provided by my hardware vendor.
Target options --> Target Architecture : arm
Target options --> Use specific FPU : vfp
Target options --> Floating point: hardware (FPU)
Target options --> Use EABI : *
Target options --> append 'hf' to the tuple (EXPERIMENTAL) : *
here is another reference  how I need to build cross tool chain
https://notsyncing.net/?p=blog&b=2016.orangepi-pc-custom-kernel

this is for your information only,please don't think otherwise, please suggest me if I am wrong.
meanwhile I will try from my end to correct myself ,if I am missing any thing.
All this is for your information,

**Comment by blprasad** [comment]
@erikd 
you can direct me at any time to make the issue to be closed status,if it's the case.

**Comment by erikd** [comment]
At the moment I don't think there is any issue here. You may have some particular issue with your toolchain or setup.

**Comment by blprasad** [comment]
@evpobr 
@erikd 

Sorry for giving you trouble, It was my mistake,I solved it by setting env #export PATH=path to arm-g++ directory.
Then every thing went smooth.When I looked at my ./configure log ,I found that arm-g++ was not picked properly,it was picking arm-gcc but did not pick arm-g++, this is what I understood so far.I don't know why it's calling g++ when the code is in C.

Anyways thanks, for your effort and time.
